src/osd/scrubber/pg_scrubber.cc

```diff
@@ -2458,117 +2401,105 @@ namespace Scrub {
-      m_waited_for_peers.push_back(p);
-      dout(10) << __func__ << ": reserve " << p.osd << dendl;
-    }
+    // send the first reservation requests
+    send_request_to_replica(*m_next_to_request, epoch);
   }
 }
 
 void ReplicaReservations::send_all_done()
 {
-  // stop any pending timeout timer
   m_osds->queue_for_scrub_granted(m_pg, scrub_prio_t::low_priority);
 }
 
 void ReplicaReservations::send_reject()
 {
-  // stop any pending timeout timer
-  m_scrub_job->resources_failure = true;
+  m_scrubber.flag_reservations_failure();
   m_osds->queue_for_scrub_denied(m_pg, scrub_prio_t::low_priority);
 }
 
-void ReplicaReservations::discard_all()
+void ReplicaReservations::release_all(replica_subset_t replicas)
 {
-  dout(10) << __func__ << ": " << m_reserved_peers << dendl;
-
-  m_had_rejections = true;  // preventing late-coming responses from triggering
-			    // events
-  m_reserved_peers.clear();
-  m_waited_for_peers.clear();
+  dout(10) << fmt::format("{}: releasing {}", __func__, replicas) << dendl;
+  epoch_t epoch = m_pg->get_osdmap_epoch();
+  // send 'release' messages to all replicas we have managed to reserve
+  for (const auto& p : replicas) {
+    release_replica(p, epoch);
+  }
 }
 
-/*
- * The following holds when update_latecomers() is called:
- * - we are still waiting for replies from some of the replicas;
- * - we might have already set a timer. If so, we should restart it.
- * - we might have received responses from 50% of the replicas.
- */
-std::optional<ReplicaReservations::tpoint_t>
-ReplicaReservations::update_latecomers(tpoint_t now_is)
+void ReplicaReservations::discard_all()
 {
-  if (m_reserved_peers.size() > m_waited_for_peers.size()) {
-    // at least half of the replicas have already responded. Time we flag
-    // latecomers.
-    return now_is + m_timeout;
-  } else {
-    return std::nullopt;
-  }
+  dout(10) << fmt::format("{}: reset w/o issuing messages", __func__) << dendl;
+  m_requests_sent = 0;
+  m_sorted_secondaries.clear();
+  m_next_to_request = m_sorted_secondaries.cbegin();
 }
 
 ReplicaReservations::~ReplicaReservations()
 {
-  m_had_rejections = true;  // preventing late-coming responses from triggering
-			    // events
-
-  // send un-reserve messages to all reserved replicas. We do not wait for
-  // answer (there wouldn't be one). Other incoming messages will be discarded
-  // on the way, by our owner.
-  epoch_t epoch = m_pg->get_osdmap_epoch();
-
-  for (auto& p : m_reserved_peers) {
-    release_replica(p, epoch);
-  }
-  m_reserved_peers.clear();
-
-  // note: the release will follow on the heels of the request. When tried
-  // otherwise, grants that followed a reject arrived after the whole scrub
-  // machine-state was reset, causing leaked reservations.
-  for (auto& p : m_waited_for_peers) {
-    release_replica(p, epoch);
-  }
-  m_waited_for_peers.clear();
+  auto reserved_by_us =
+      replica_subset_t{m_sorted_secondaries.cbegin(), m_requests_sent};
```

Comment:
Let's just use the existing m_next_to_request iterator.  When you get a denial, decrement the iterator.  You can derive the number of requests sent by m_next_to_request - m_sorted_secondaries.cbegin() -- num_requests_sent().